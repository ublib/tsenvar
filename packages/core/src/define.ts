// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Defs = { [key: string]: Def<any> };

export interface Def<T extends EnvVarConstructor> {
  /**
   * constructor with validation.
   *
   * when throw error, process exits with code 1.
   *
   * @example
   *
   * ```ts
   * // built-in type
   * construct: String
   * construct: RegExp
   * ```
   * ```ts
   * // custom type and validator
   * construct: (v: any): 'a' | 'b' | 'c' => {
   *   if (!['a', 'b', 'c'].includes(v)) throw new EnvVarLib.InvalidError('unexpected NEW_VAR')
   *   return v
   * }
   * ```
   *
   */
  construct: T;

  /** default: `false` */
  optional?: boolean;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type EnvVarConstructor = (...args: any) => any;

export type ResolveType<T> = { [K in keyof T]: ResolveEnvVarType<T[K]> };

type ResolveEnvVarType<T> = T extends { construct: infer U }
  ? U extends EnvVarConstructor
    ? ReturnType<U>
    : never
  : never;

// for type inference
export const define = <T extends Defs>(def: T): T => def;
