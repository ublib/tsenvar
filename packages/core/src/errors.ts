export type ErrorOnLoad = DuplicateError | ValidationError;

export class ValidationError extends Error {
  constructor(public readonly reason: string) {
    super(reason);
  }
}

export class DuplicateError extends Error {
  constructor(public readonly reason: string) {
    super(reason);
  }
}
