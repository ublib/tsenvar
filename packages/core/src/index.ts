import type { Defs, ResolveType } from "./define";

export * from "./errors";

export type { Defs, Def, ResolveType } from "./define";
export { define } from "./define";

export interface Tsenvar {
  load: Load;
  clear: Clear;
}

export interface LoaderOptions {
  /** @default ".env" */
  envPath?: string;

  /** @default false */
  overwrite?: boolean;
}

export type Load = <T extends Defs>(envVarDefs: T, options?: LoaderOptions) => ResolveType<T>;

export type Clear = (defs: Defs) => void;

export * from "./internal";
