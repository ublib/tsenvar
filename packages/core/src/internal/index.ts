import dotenv from "dotenv";

import type { Tsenvar } from "..";
import type { LoaderContext } from "./interface";

import { clearer } from "./clear";
import { mergeLoaderOptions } from "./options";
import { resolve } from "./resolve";

export * from "./interface";

/**
 * @internal
 */
export const createTsenvar = (ctx: Readonly<LoaderContext>): Tsenvar => {
  return { load: loader(ctx), clear: clearer(ctx) };
};

const loader =
  (ctx: Readonly<LoaderContext>): Tsenvar["load"] =>
  (envVarDefs, options) => {
    const mergedOptions = mergeLoaderOptions(options);
    // TODO: load any runtime
    dotenv.config({ path: mergedOptions.envPath });
    return resolve(ctx, envVarDefs);
  };
