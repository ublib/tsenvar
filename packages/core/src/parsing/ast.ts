export interface Span {
  start: number;
  end: number;
}

export interface Node {
  span: Span;
}

export interface Document extends Node {
  envVars: EnvVar[];
}

export interface EnvVar extends Node {
  id: Identifier;
  value: Value;
}

export interface Identifier extends Node {
  name: string;
}

export interface Value extends Node {
  value: string;
}
