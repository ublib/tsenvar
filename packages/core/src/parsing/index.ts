export type { Node, EnvVar, Identifier, Value } from "./ast";
export { parse } from "./parser";
