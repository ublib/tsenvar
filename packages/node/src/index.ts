import fs from "node:fs";

import type { LoaderContext } from "@tsenvar/core";
import { createTsenvar } from "@tsenvar/core";

const loaderContext = {
  rawEnv: {
    read: () => process.env,
    write: (k, v) => {
      process.env[k] = v;
    },
  },

  panic: (reason, status) => {
    process.stderr.write(reason);
    process.exit(status ?? 1);
  },

  fs: {
    readFileSync: (path) => fs.readFileSync(path, "utf-8"),
  },
} satisfies LoaderContext;

export const Tsenvar = createTsenvar(loaderContext);
