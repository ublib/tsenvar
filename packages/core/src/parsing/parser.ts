import { EnvVar, Document, Identifier, Value, Node } from "./ast";

const EOF = "\0";

type ParseResult<T extends Node> = ParseResultOk<T> | ParseResultErr;

interface ParseResultBase {
  ok: boolean;
}

interface ParseResultOk<T extends Node> extends ParseResultBase {
  ok: true;
  value: T;
}

interface ParseResultErr extends ParseResultBase {
  ok: false;
  error: string;
}

interface ParseErrorValue {
  error: string;
  position: number;
}

const ok = <T extends Node>(value: T): ParseResult<T> => ({
  ok: true,
  value,
});

interface ParserContext {
  source: string;
  position: number;
}

const createContext = (source: string): ParserContext => ({ source, position: 0 });

export const parse = (input: string): ParseResult<Document> => {
  const context = createContext(input);
  return parseDocument(context);
};

const parseDocument = (context: ParserContext): ParseResult<Document> => {
  const envVars: EnvVar[] = [];
  const errors: ParseErrorValue[] = [];
  while (peekChar(context) !== EOF) {
    const envVar = parseEnvVar(context);
    if (!envVar.ok) {
      errors.push({ error: envVar.error, position: context.position });
      continue;
    }
    envVars.push(envVar.value);
  }
  return ok({
    envVars,
    span: { start: 0, end: context.position },
  });
};

const parseEnvVar = (context: ParserContext): ParseResult<EnvVar> => {
  const start = context.position;
  const id = parseIdentifier(context);
  if (!id.ok) {
    return id;
  }
  consumeWhitespace(context);
  consumeChar(context);
  consumeWhitespace(context);
  const value = parseValue(context);
  if (!value.ok) {
    return value;
  }
  const end = context.position;
  return ok({
    id: id.value,
    value: value.value,
    span: { start, end },
  });
};

const parseIdentifier = (context: ParserContext): ParseResult<Identifier> => {
  const start = context.position;
  let name = "";
  while (peekChar(context) !== "=") {
    name += consumeChar(context);
  }
  const end = context.position;
  return ok({
    name,
    span: { start, end },
  });
};

const parseValue = (context: ParserContext): ParseResult<Value> => {
  const start = context.position;
  let value = "";
  while (peekChar(context) !== "\n") {
    value += consumeChar(context);
  }
  const end = context.position;
  return ok({
    value,
    span: { start, end },
  });
};

const consumeChar = (context: ParserContext): string => {
  return context.source[context.position++] ?? EOF;
};

const consumeWhitespace = (context: ParserContext): void => {
  while (/s/.test(peekChar(context))) {
    consumeChar(context);
  }
};

const peekChar = (context: ParserContext): string => {
  return context.source[context.position] ?? EOF;
};
