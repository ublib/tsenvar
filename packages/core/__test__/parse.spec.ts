import { describe, it, expect } from "vitest";
import { parse } from "../src/parsing";

describe("parse", () => {
  it("should parse", () => {
    const result = parse("STAGE_NAME=");
    expect(result.ok).toBe(true);
    if (result.ok) {
      expect(result.value).toEqual({
        envVars: [
          {
            id: {
              name: "STAGE_NAME",
              span: { start: 0, end: 10 },
            },
            value: {
              value: "",
              span: { start: 10, end: 10 },
            },
            span: { start: 0, end: 11 },
          },
        ],
        span: { start: 0, end: 11 },
      });
    }
  });

  it("should parse with space", () => {
    const result = parse(`STAGE_NAME
      =`);
    expect(result.ok).toBe(true);
    if (result.ok) {
      expect(result.value).toEqual({
        envVars: [
          {
            id: {
              name: "STAGE_NAME",
              span: { start: 0, end: 10 },
            },
            value: {
              value: "",
              span: { start: 10, end: 10 },
            },
            span: { start: 0, end: 18 },
          },
        ],
        span: { start: 0, end: 18 },
      });
    }
  });
});
