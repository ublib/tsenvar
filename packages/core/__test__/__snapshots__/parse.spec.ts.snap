// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`success to multi vars parse > should parse 1`] = `
{
  "envVars": [
    {
      "id": {
        "name": "STAGE_NAME",
        "span": {
          "end": 10,
          "start": 0,
        },
      },
      "span": {
        "end": 22,
        "start": 0,
      },
      "value": {
        "span": {
          "end": 22,
          "start": 11,
        },
        "value": "development",
      },
    },
    {
      "id": {
        "name": "PORT",
        "span": {
          "end": 27,
          "start": 23,
        },
      },
      "span": {
        "end": 32,
        "start": 23,
      },
      "value": {
        "span": {
          "end": 32,
          "start": 28,
        },
        "value": "3000",
      },
    },
  ],
  "span": {
    "end": 32,
    "start": 0,
  },
}
`;

exports[`success to multi vars parse > should parse mixed double quat and no double quot values 1`] = `
{
  "envVars": [
    {
      "id": {
        "name": "STAGE_NAME",
        "span": {
          "end": 10,
          "start": 0,
        },
      },
      "span": {
        "end": 24,
        "start": 0,
      },
      "value": {
        "span": {
          "end": 23,
          "start": 12,
        },
        "value": "development",
      },
    },
    {
      "id": {
        "name": "PORT",
        "span": {
          "end": 31,
          "start": 27,
        },
      },
      "span": {
        "end": 36,
        "start": 27,
      },
      "value": {
        "span": {
          "end": 36,
          "start": 32,
        },
        "value": "3000",
      },
    },
  ],
  "span": {
    "end": 36,
    "start": 0,
  },
}
`;

exports[`success to multi vars parse > should parse mixed double quat and no double quot values 2`] = `
{
  "envVars": [
    {
      "id": {
        "name": "STAGE_NAME",
        "span": {
          "end": 10,
          "start": 0,
        },
      },
      "span": {
        "end": 22,
        "start": 0,
      },
      "value": {
        "span": {
          "end": 22,
          "start": 11,
        },
        "value": "development",
      },
    },
    {
      "id": {
        "name": "PORT",
        "span": {
          "end": 29,
          "start": 25,
        },
      },
      "span": {
        "end": 36,
        "start": 25,
      },
      "value": {
        "span": {
          "end": 35,
          "start": 31,
        },
        "value": "3000",
      },
    },
  ],
  "span": {
    "end": 36,
    "start": 0,
  },
}
`;

exports[`success to multi vars parse > should parse mixed double quat and no double quot values 3`] = `
{
  "envVars": [
    {
      "id": {
        "name": "STAGE_NAME",
        "span": {
          "end": 10,
          "start": 0,
        },
      },
      "span": {
        "end": 28,
        "start": 0,
      },
      "value": {
        "span": {
          "end": 28,
          "start": 17,
        },
        "value": "development",
      },
    },
    {
      "id": {
        "name": "PORT",
        "span": {
          "end": 35,
          "start": 31,
        },
      },
      "span": {
        "end": 46,
        "start": 31,
      },
      "value": {
        "span": {
          "end": 45,
          "start": 41,
        },
        "value": "3000",
      },
    },
  ],
  "span": {
    "end": 46,
    "start": 0,
  },
}
`;

exports[`success to multi vars parse > should parse with double quat value 1`] = `
{
  "envVars": [
    {
      "id": {
        "name": "STAGE_NAME",
        "span": {
          "end": 15,
          "start": 5,
        },
      },
      "span": {
        "end": 29,
        "start": 5,
      },
      "value": {
        "span": {
          "end": 28,
          "start": 17,
        },
        "value": "development",
      },
    },
    {
      "id": {
        "name": "PORT",
        "span": {
          "end": 34,
          "start": 30,
        },
      },
      "span": {
        "end": 41,
        "start": 30,
      },
      "value": {
        "span": {
          "end": 40,
          "start": 36,
        },
        "value": "3000",
      },
    },
    {
      "id": {
        "name": "PORT2",
        "span": {
          "end": 47,
          "start": 42,
        },
      },
      "span": {
        "end": 56,
        "start": 42,
      },
      "value": {
        "span": {
          "end": 55,
          "start": 51,
        },
        "value": "3000",
      },
    },
  ],
  "span": {
    "end": 57,
    "start": 0,
  },
}
`;

exports[`success to multi vars parse > should parse with double quat value with spaces 1`] = `
{
  "envVars": [
    {
      "id": {
        "name": "STAGE_NAME",
        "span": {
          "end": 15,
          "start": 5,
        },
      },
      "span": {
        "end": 35,
        "start": 5,
      },
      "value": {
        "span": {
          "end": 34,
          "start": 17,
        },
        "value": "
    development
",
      },
    },
    {
      "id": {
        "name": "PORT",
        "span": {
          "end": 45,
          "start": 41,
        },
      },
      "span": {
        "end": 52,
        "start": 41,
      },
      "value": {
        "span": {
          "end": 51,
          "start": 47,
        },
        "value": "3000",
      },
    },
    {
      "id": {
        "name": "PORT2",
        "span": {
          "end": 59,
          "start": 54,
        },
      },
      "span": {
        "end": 69,
        "start": 54,
      },
      "value": {
        "span": {
          "end": 68,
          "start": 63,
        },
        "value": "
3000",
      },
    },
  ],
  "span": {
    "end": 70,
    "start": 0,
  },
}
`;

exports[`success to multi vars parse > should parse with space head 1`] = `
{
  "envVars": [
    {
      "id": {
        "name": "STAGE_NAME",
        "span": {
          "end": 15,
          "start": 5,
        },
      },
      "span": {
        "end": 27,
        "start": 5,
      },
      "value": {
        "span": {
          "end": 27,
          "start": 16,
        },
        "value": "development",
      },
    },
    {
      "id": {
        "name": "PORT",
        "span": {
          "end": 32,
          "start": 28,
        },
      },
      "span": {
        "end": 37,
        "start": 28,
      },
      "value": {
        "span": {
          "end": 37,
          "start": 33,
        },
        "value": "3000",
      },
    },
  ],
  "span": {
    "end": 37,
    "start": 0,
  },
}
`;
